Load data to child UI

AryamanDubey28 opened the pull request:
Hi guys,

For my task I had to load data to the child UI. This was similar to Ali's task and I used a class made by him (FirebaseFunctions) and added minimal edits to it as well. I added some tests for this task and my Sprint 2 task as well. I would appreciate it if you guys could lmk about any thoughts/changes!

Thanks,
Aryaman

Commit:
show data from dev categories in child ui

Commit:
change network image to clickable image

Commit:
show data from deCategories list

Commit:
add listener features using stream builder and refreshing state

Commit:
fix failing tests

Commit:
fix failing tests

Commit:
Merge branch 'master' into load-data-to-child-ui

Commit:
add more PIN tests

Commit:
add more tests

Commit:
fix error

Commit:
add test fields

Commit:
add db test

Commit:
remove unused code

Commit:
add functionality that if item is unavailable, it is not shown

Commit:
unavailable images still show on choice board

Commit:
add firebasefirestore to auth

Commit:
add firebase firestore to auth

Commit:
make getUserPIN test

Commit:
make getCurrentUserId test

Commit:
remove unused code and add child logout widget test

Commit:
remove comment

Commit:
Merge branch 'master' into load-data-to-child-ui

Commit:
merge master into branch

Commit:
merge master

Commit:
merge master

Commit:
fix failing tests

Commit:
fix ios build issue?

Commit:
fix 2?

Commit:
fix pod issue try 3

dmunipi1 reviewed (approved):
Good job Aryaman, everything looks good

absirgue commented on lib/main.dart:
```diff
@@ -20,7 +21,8 @@ Future main() async {
 
   final pref = await SharedPreferences.getInstance();
-  final auth = Auth(auth: FirebaseAuth.instance);
+  final auth =
```
Here, make the change related to comment #2

absirgue commented on lib/pages/authenticate/edit_account.dart:
```diff
@@ -27,13 +35,28 @@ class EditAccountPageState extends State<EditAccountPage> {
+  }
+
+  // Displays an alert dialog with the text passed as parameter.
+  void show_alert_dialog(String text) {
```
delete this method (this probably indicates that merge conflicts were not solved the right way because I did remove that one. Please make sure that all the places that show an alert dialog use the helper class I built (error_dialog_helper.dart)

absirgue commented on lib/pages/authenticate/login.dart:
```diff
@@ -29,7 +30,9 @@ class _LogInState extends State<LogIn> {
-    auth = Auth(auth: widget.auth);
+    FirebaseAuth firebaseAuth = FirebaseAuth.instance;
+    FirebaseFirestore firebaseFirestore = FirebaseFirestore.instance;
+    auth = Auth(auth: firebaseAuth, firestore: firebaseFirestore);
```
here as well, please make change related to comment #2

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -89,15 +97,18 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
   Future<void> submit(BuildContext context) async {
```
Please rename this method to something more meaningful

absirgue replied:
Please also try to divide this method in multiple smaller ones as it is very long

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -113,47 +124,121 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
-                //only triggers when its pressed for some time and swiped up
-                onLongPressUp: () async {
-                  openLogoutDialog(context);
+    if (!widget.mock) {
```
Please but all the code in the if statement in a stand alone method

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -1,68 +1,76 @@
-//
 
 class CustomizableColumn extends StatefulWidget {
   final bool mock;
```
please rename to isMock

absirgue replied:
still stands

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -1,68 +1,76 @@
   final bool mock;
+  static int customizableColumnRequestCounter = 0; //used for testing
+  List<List<ClickableImage>> testList; //used for testing
+  CustomizableColumn({this.mock = false, testList})
```
why is the field here called testList. Is it used for testing only? Otherwise please rename

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -113,47 +124,121 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
-        context,
-        MaterialPageRoute(builder: (context) => AdminChoiceBoards()),
-      );
+      if (pin_controller.text.trim() == "0000") {
```
why is the pin "0000" hardcoded here?

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -113,47 +124,121 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
+                    storage: mockFirebaseStorage,
+                  )),
+        );
+        print("done");
```
please remove all print statements

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -113,47 +124,121 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
+          },
+        ),
+      );
+    } else {
```
just realized how large this if statement is. Please do not do that, if the view you are testing is 100% different from the displayed view then there isn't much point to testing it. Please refactor the code to use essentially the same view for both testing and normal purposes. The differences should happen either by mocking the connection to the database or by not doing the database calls and passing in a fake list. But the build() method can not be that different for testing and normal situations, it defeats the purpose

absirgue commented on lib/services/admin.dart:
```diff
@@ -0,0 +1,38 @@
+
+  Future<String?> getCurrentUserId() async {
+    //var user = await auth.currentUser;
+    if (!mock) {
```
we can not have al the methods use as isMock boolean. That is why we use mocking. Please refactor this method and do appropriate changes in the tests to properly mock a user and for this method to return a fake userId not bcs it's hardcoded in it and it knows it is operating as a test but because we pass it a fake FirebaseAuth object and a fake User object that has this id

absirgue commented on lib/services/auth.dart:
```diff
@@ -50,23 +52,35 @@ class Auth {
   }
 
   Future<bool> checkPINExists() async {
```
please use lower case for PIN

absirgue replied:
I mean change PIN to Pin

absirgue commented on lib/services/auth.dart:
```diff
@@ -50,23 +52,35 @@ class Auth {
-            'userId': await getCurrentUserId(),
-          };
-          await docUser.set(entry); //adds PIN to database
+          if (!mock) {
```
why checking if not mock here. Again, no method should check for that. The only place in the code where this is possible is to prevent a loading indicator from appearing during a test. Please remove ALL the if (isMock) in your code and replace them by proper mocking in your tests

absirgue commented on lib/services/auth.dart:
```diff
@@ -124,27 +138,31 @@ class Auth {
-            .then((value) {
-          value.docs.forEach((element) {
-            docId = element.id;
+        if (!mock) {
```
same here

absirgue commented on test/screens/auth/edit_account_test.dart:
```diff
@@ -0,0 +1,198 @@
+import 'dart:math';
```
I added test for edit_account as well so there should be one file called edit_account_widget_test and the other edit_account_unit_test. Please put your tests in the proper one.  the edit_account_test file must be deleted

absirgue reviewed (changes requested):
Hey,

Thank for the work.

I have a few comments, please address them asap bcs tonight was the deadline for this task whose deadline was last Monday.

Here are the comments:
1- Please check the two screenshots attached to identify parts of the code that are not tested and for which you should add some tests.
2- Please refactor all related to the change to the Auth class. You don't need to make the firestore and auth (change name maybe as well bcs it's not very clear but that's not your fault - you can rename the variable to "firebaseAuth") required, you can have them optional (normal fields) and add a body to the constructor that says that if the passed in values are not null then make the fields have this value and if they are null then make it use FirebaseAuth .instance and Firestore.instance. (you can check a lot of of other classes that do the same e.g. the edit account class). And do the relevant changes to delete everywhere where you set the value of these parameters to be Firestore.instance or FirebaseAuth.instance. This makes for a shorter code and the "normal" classes should not suffer from our tests' requirements to be able to pass in Mock classes if you see what I mean. It will also make your code more consistent with the rest of the program.

3- Check the in-code comments below.

absirgue commented:
<img width="823" alt="Screen Shot 2023-03-19 at 9 24 07 PM" src="https://user-images.githubusercontent.com/94190429/226212136-81d70ce2-159c-4e7f-8c50-c1581fba7b1e.png">
<img width="926" alt="Screen Shot 2023-03-19 at 9 24 23 PM" src="https://user-images.githubusercontent.com/94190429/226212140-023ffd22-dcb3-43b1-8b05-ff1f6f05dd72.png">
<img width="1264" alt="Screen Shot 2023-03-19 at 9 24 36 PM" src="https://user-images.githubusercontent.com/94190429/226212141-28e6eb61-3fbe-4ffd-9e66-6007ddbb63f2.png">

absirgue commented:
As said on the group chat, I discovered some issues in the functionality itself.

1- when I mark a category unavailable, the categoryItem that was in it shows in the place of the category and its cover picture. 
2- when I add an item to a category it does not display in the menu. Actually it does in normal circonstances but I had a case where it did not update properly. Sadly I don't have much more info for you to reproduce the bug. Please just check for the robustness of your code. (see screenshot to see the problem)

<img width="681" alt="Screen Shot 2023-03-19 at 9 53 29 PM" src="https://user-images.githubusercontent.com/94190429/226212496-6a3f02b4-c2dc-43cc-ad80-4da5165da332.png">
<img width="730" alt="Screen Shot 2023-03-19 at 9 53 35 PM" src="https://user-images.githubusercontent.com/94190429/226212498-0985de86-3873-4c0e-9869-fff5aa2440b6.png">

3- categories with no available items still show. You can ask on the group chat but my understanding from last meeting is that you can go ahead and make them not show to improve the UX.

Commit:
make changes and fix failing tests

Commit:
add test for filtered images

Commit:
add tests, fix failing test and refactor Auth

Ali-Alkhars commented on lib/pages/admin/admin_side_menu.dart:
```diff
@@ -95,26 +98,45 @@ class AdminSideMenu extends StatelessWidget {
+                if (check) {
+                  //PIN exists
+
+                  // final pref = await SharedPreferences.getInstance();
```
Remove commented out code

Ali-Alkhars commented on lib/pages/child_board/child_board.dart:
Remove unused imports in this file

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -1,68 +1,93 @@
+    completer = Completer();
+    buildCompleter();
+    authentitcationHelper = Auth(auth: auth, firestore: firebaseFirestore);
+    timer = Timer.periodic(Duration(seconds: 4), (timer) {
```
Make it 30 sec (and update the comment)

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -81,49 +106,54 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
+    String currentPin = await authentitcationHelper.getCurrentUserPIN();
+
+    if (pin_controller.text.trim() == currentPin) {
+      // final pref = await SharedPreferences.getInstance();
```
remove commented out code

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -81,49 +106,54 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
   // Construct a column of rows using category title and images
   @override
   Widget build(BuildContext context) {
+    //buildCompleter();
```
remove commented out code

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -133,25 +163,54 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
+            ? getListFromChoiceBoards()
+            : completer
+                .future, //retrieves a list from the database of categories and items associated with the category
+
```
remove commented out code

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
Remove unused imports

Ali-Alkhars commented on lib/pages/child_menu/customizable_row.dart:
```diff
@@ -31,20 +44,34 @@ class CustomizableRow extends StatelessWidget {
+          List<ClickableImage> newList = widget.unfilteredImages
+              .where((x) => widget.unfilteredImages.indexOf(x) != 0)
+              .toList();
           // PLACEHOLDER.
```
Delete `// PLACEHOLDER` comment

Commit:
refactor logout button into its own branch

Ali-Alkhars commented on lib/services/auth.dart:
```diff
@@ -49,18 +54,23 @@ class Auth {
         if (pin.length == 4 && num.tryParse(pin) != null) {
-          final docUser =
-              FirebaseFirestore.instance.collection('userPins').doc();
+          // final docUser =
```
remove commented out code

Ali-Alkhars commented on lib/widgets/logout_icon_button.dart:
```diff
@@ -0,0 +1,99 @@
+    String currentPin = await authenticationHelper.getCurrentUserPIN();
+
+    if (pin_controller.text.trim() == currentPin) {
+      // final pref = await SharedPreferences.getInstance();
```
remove commented out code

Ali-Alkhars commented on lib/widgets/logout_icon_button.dart:
```diff
@@ -0,0 +1,99 @@
+import '../pages/admin/admin_choice_boards.dart';
+import '../services/auth.dart';
+
+class LogoutIconButton extends StatelessWidget {
```
Add comment

Ali-Alkhars commented on lib/widgets/logout_icon_button.dart:
```diff
@@ -0,0 +1,99 @@
+    this.authenticationHelper = authenticationHelper;
+  }
+
+  Future openLogoutDialog(BuildContext context) => showDialog(
```
Add comment

Ali-Alkhars commented on lib/widgets/logout_icon_button.dart:
```diff
@@ -0,0 +1,99 @@
+            ],
+          ));
+
+  Future<void> submitPin(BuildContext context) async {
```
Add comment

Ali-Alkhars commented on test/screens/admin/admin_side_menu_test.dart:
Remove unused imports

Ali-Alkhars commented on test/screens/child/child_board_test.dart:
```diff
@@ -1,29 +1,47 @@
+      ClickableImage test_image = ClickableImage(
+          name: "test",
+          imageUrl:
+              "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAoHCBUVFBcVFRUYGBcYGxobHBsbGBsbGh0gGx0bHCAcHh0bICwkICArIhobJTYlKS4wMzMzHSI5PjkxPSwyMzABCwsLEA4QHhISHTQqIik0MjsyMjIyMjIyNDIyMjIyMjIyMjIyMjIyNDIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMv/AABEIAOEA4QMBIgACEQEDEQH/xAAcAAACAgMBAQAAAAAAAAAAAAAFBgMEAAIHAQj/xABGEAABAgMFBgQEAwUGBQQDAAABAhEAAyEEBRIxQQYiUWFxgRMykaFSscHwQtHhFCNygpIVU2Ky0vEzNEOiwgcWc ...
```
Please change the image!!!!!

Ali-Alkhars commented on test/screens/child/child_board_test.dart:
```diff
@@ -1,29 +1,47 @@
+            ),
           )));
 
       // Verify that the board menu is displayed
```
You don't really need comments here

Ali-Alkhars commented on test/screens/child/child_menu_test.dart:
Delete unused imports

Ali-Alkhars commented on lib/models/clickable_image.dart:
```diff
@@ -7,4 +9,13 @@ class ClickableImage {
   ClickableImage(
       {required this.name, required this.imageUrl, required this.is_available});
+
+  factory ClickableImage.fromSnapshot(
```
You can remove this

Ali-Alkhars commented on lib/pages/admin/admin_side_menu.dart:
- Remove unused imports
- Remove `show_alert_dialog` method

Ali-Alkhars reviewed (changes requested):
Good work! You just need to enhance code quality (check comments below) and:
- Add tests for `widgets/edit_pin_section.dart`

Commit:
enhance code quality by removing comments and uneeded imports

Commit:
improve code quality

Commit:
add some edit PIN tests

Commit:
add edit PIN tests

Commit:
remove print statements

Commit:
remove print

Commit:
Merge branch 'master' into load-data-to-child-ui

Commit:
changes

Commit:
resolve merge conflicts

Commit:
resolve merge conflicts

Commit:
make tests to improve test coverage

Commit:
make edit pin test

absirgue commented on lib/data/choice_boards_data.dart:
```diff
@@ -1,11 +1,201 @@
+//             name: "Butter chicken",
+//             imageUrl:
+//                 "https://www.cookingclassy.com/wp-content/uploads/2021/01/butter-chicken-4.jpg"),
+//         CategoryItem(
```
delete all commented out code

absirgue commented on lib/pages/admin/admin_side_menu.dart:
```diff
@@ -40,6 +43,11 @@ class AdminSideMenu extends StatelessWidget {
 
+  // Displays an alert dialog with the text passed as parameter.
+  // void show_alert_dialog(BuildContext context, String text) {
+  //   ShowAlertDialog.show_dialog(context, text);
```
delete all commented out code

absirgue commented on lib/pages/authenticate/login.dart:
```diff
@@ -8,34 +10,43 @@ import 'package:seg_coursework_app/widgets/my_text_field.dart';
     super.initState();
-    auth = Auth(auth: widget.auth);
+
+    authentitcationHelper = Auth(auth: auth, firestore: firebaseFirestore);
```
here, the login class should have to have any code related to Firestore, only FirebaseAuth. Please either of two things:
- move all the pin related methods to the firestore_functions (bcs they are not linked with firebaseAuth)
- make the firestore parameter of Auth class defaults to be Firestore.instance and make it optional (like u did for this class) so that the classes that use normal FirebaseAuth do not have to worry about Firestore


Option 1 is the best one but I let you decide if it makes sense.

absirgue commented on lib/pages/child_board/child_board.dart:
```diff
@@ -145,12 +94,31 @@ class _ChildBoards extends State<ChildBoards> with TickerProviderStateMixin {
     );
   }
 
+  // Returns category image that shows on menu
```
make this comment a bit more clear please: "Returns the cover picture of the category on display" something like that

absirgue commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -131,27 +87,45 @@ class _CustomizableColumnState extends State<CustomizableColumn> {
-          return const Divider(height: 2);
+      body: FutureBuilder(
+        future: !widget
+                .mock //getListFromChoiceBoards updates from Firestore but when testing it will cause tester.pumpAndSettle() to time out, completer.future will allow tests to work perfectly but won't allow to have updates from the db. To achieve the best of both worlds, both are here in a ternary operator
```
make comment on its own line bcs it's pretty big

absirgue commented on lib/pages/child_menu/customizable_row.dart:
```diff
@@ -41,21 +52,33 @@ class CustomizableRow extends StatelessWidget {
+                  builder: (context) => ChildBoards(
+                        categoryTitle: widget.categoryTitle,
+                        categoryImage: widget.imagePreviews[
+                            0], //first image forms image of category
```
call it "cover picture of the category" in the comment

absirgue reviewed (changes requested):
Hey,

I left a few comments in the code itself.

Apart from this, please add the test we discussed for successful making of a pin. In addition, here a few parts of the code that are not tested:

(Checking that logging out doesn't work if PIN is not correct)
<img width="700" alt="Screen Shot 2023-03-21 at 2 04 33 PM" src="https://user-images.githubusercontent.com/94190429/226631094-9084b7ac-a6f1-413b-a484-efba19ec7fb7.png">


(Making items unavailable and them displaying as such)
<img width="774" alt="Screen Shot 2023-03-21 at 2 06 09 PM" src="https://user-images.githubusercontent.com/94190429/226631421-53fccb5c-f9f4-4333-a875-cfaa25116981.png">

Thanks!

Commit:
add changes requested on pr

absirgue committed:
Remove unused method

absirgue committed:
Restructure child_board method

absirgue committed:
Pull main and solve conflicts

absirgue committed:
bring back items marked with red cross when unavailable

absirgue committed:
Use team models and change code structure

absirgue committed:
Fix tests for new code structure

absirgue committed:
Arrange file structure

absirgue committed:
Adjust tests for new file structure

absirgue committed:
Make edit account page reload on edits

absirgue committed:
Add edit account navigation test

Ali-Alkhars committed:
General refactoring

- Remove commented-out code
- Add missing comments
- Rename some variables
- Add const

skellyhoe committed:
Make cross bigger on child view. Show less images on portrait mode.
